Disable keyscan on `git://` urls.

dochang opened the pull request:
`git://` isn't ssh url.  Do not check the host key of it.

abadger commented:
I don't think we want to change get_fqdn() in this way.  get_fqdn() is just a helper function that gets a fully qualified domain name from a uri.  Disabling keyscan for git:// is more of a policy decision and should be tied to a specific action rather than the helper function.

Mentioning a reproducer for the problem you're having might make it easier to see what area of the code the fix belongs to.

dochang commented:
`get_fqdn` gets an FQDN from a **[giturl](/ansible/ansible/blob/38b0a6ae875473c00a40e2b6b9b002ec5216a29d/lib/ansible/module_utils/known_hosts.py#L57)** when it's based on **[ssh](/ansible/ansible/blob/38b0a6ae875473c00a40e2b6b9b002ec5216a29d/lib/ansible/module_utils/known_hosts.py#L73)**.

But `git://` is also considered as an ssh url by mistake.  It should not be a policy decision.

To reproduce it, setup a git daemon on a server whose host key is **NOT** in your "known_hosts" files (`~/.ssh/known_hosts`, `/etc/ssh/ssh_known_hosts`, etc.).  Then use ansible to clone the git repo `git://githost/repo.git`.  You will get [this error](/ansible/ansible/blob/38b0a6ae875473c00a40e2b6b9b002ec5216a29d/lib/ansible/module_utils/known_hosts.py#L53).

abadger commented:
<nod>.  So judging from where the error is coming from, the change should be in add_git_host_key().  That's what I mean by policy.  get_fqdn() should just retrieve the FQDN from a URI.  add_git_host_key() should decide that a given uri doesn't need to have a host key added for it and act appropriately.

dochang committed:
Add host key for ssh url only.

Rewrite function `get_fqdn`.  It returns fqdn for all kinds of urls now.

`add_git_host_key` determines whether a url is ssh and whether its host
key should be added.

dochang commented:
Understand you.  I've pushed a new patch.  Please check it again.

kportertx commented:
@dochang getfqdn doesn't honor socket timeouts and can take about 40 seconds to timeout. I normally work around this by using something similar to the following:

```
def getfqdn(address, timeout=0.5):
    # note: cannot use timeout lib because signal must be run from the
    #       main thread
    result = [address]

    def helper():
        result[0] = socket.getfqdn(address)

    t = threading.Thread(target=helper)
    t.start()
    t.join(timeout)
    return result[0]
```

dochang commented:
@kportertx `get_fqdn` in `known_hosts.py` and `getfqdn` in `facts.py` are different.

`get_fqdn` only parses the url string and returns the hostname.  There're no socket operations in it.

amenonsen commented:
This PR still seems relevant and desirable and seems reasonable. @dochang, do you feel like rebasing this onto devel?

dochang commented:
Done. @amenonsen

abadger commented:
Sorry I lost track of this one.  Looks like a good change.  i'll massage the commit so that it applies cleanly and merge.

abadger commented:
Merged: https://github.com/ansible/ansible/commit/58c3539196a0a44876913a7cc54ca6dd5c1ab96a

Thanks @dochang!  Sorry for the long wait to get this in.

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
